shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/route/engine/validator/dml/impl/ShardingSelectStatementValidator.java

```diff
@@ -40,6 +41,11 @@ public void preValidate(final ShardingRule shardingRule, final SQLStatementConte
         if (isNeedMergeShardingValues(sqlStatementContext, shardingRule)) {
             needCheckDatabaseInstance = checkSubqueryShardingValues(shardingRule, sqlStatementContext, parameters, schema);
         }
+        if (!sqlStatementContext.getSqlStatement().getUnionSegments().isEmpty()) {
```

Comment:
@tuichenchuxin Please extract this logic to a method, and remove `isAllSingleTables` method.